go/tasks/plugins/array/k8s/task.go

```diff
@@ -103,22 +104,36 @@ func (t Task) Launch(ctx context.Context, tCtx core.TaskExecutionContext, kubeCl
 	if t.Config.RemoteClusterConfig.Enabled {
 		podTemplate.OwnerReferences = nil
 	}
+
 	if len(podTemplate.Spec.Containers) == 0 {
 		return LaunchError, errors2.Wrapf(ErrReplaceCmdTemplate, err, "No containers found in podSpec.")
 	}
+
 	containerIndex, err := getTaskContainerIndex(&podTemplate)
 	if err != nil {
 		return LaunchError, err
 	}
 
 	indexStr := strconv.Itoa(t.ChildIdx)
 	podName := formatSubTaskName(ctx, tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName(), indexStr)
+	allocationStatus, err := allocateResource(ctx, tCtx, t.Config, podName)
+	if err != nil {
+		return LaunchError, err
+	}
+
+	if allocationStatus != core.AllocationStatusGranted {
+		t.NewArrayStatus.Detailed.SetItem(t.ChildIdx, bitarray.Item(core.PhaseWaitingForResources))
+		t.NewArrayStatus.Summary.Inc(core.PhaseWaitingForResources)
+		return LaunchWaiting, nil
+	}
 
 	pod := podTemplate.DeepCopy()
 	pod.Name = podName
 	pod.Spec.Containers[containerIndex].Env = append(pod.Spec.Containers[containerIndex].Env, corev1.EnvVar{
-		Name:  FlyteK8sArrayIndexVarName,
-		Value: indexStr,
+		Name: FlyteK8sArrayIndexVarName,
+		// Use the OriginalIndex which represents the position of the subtask in the original user's map task before
+		// compacting indexes caused by catalog-cache-check.
+		Value: strconv.Itoa(t.OriginalIndex),
```

Comment:
woah, was this just horribly broken before?

Reply:
yup... :(